h/services/rename_user.py

```diff
@@ -42,26 +42,34 @@ def rename(self, user, new_username):
         self.check(user, new_username)
 
         old_userid = user.userid
-
         user.username = new_username
         new_userid = user.userid
 
+        # Remove auth tickets when renaming the user. We cannot just update the
+        # denormalized `user_userid` of these because the previous userid values
+        # will have been serialized into the session cookies stored in the
+        # user's browser. See
+        # https://michael.merickel.org/projects/pyramid_auth_demo/auth_vs_auth.html
         self._purge_auth_tickets(user)
-        self._purge_tokens(user)
 
-        ids = self._change_annotations(old_userid, new_userid)
+        # For OAuth tokens, only the token's value is stored by clients, so we
+        # can just update the userid.
+        self._update_tokens(old_userid, new_userid)
 
+        ids = self._change_annotations(old_userid, new_userid)
         self.reindex(ids)
 
     def _purge_auth_tickets(self, user):
         self.session.query(models.AuthTicket) \
             .filter(models.AuthTicket.user_id == user.id) \
             .delete()
 
-    def _purge_tokens(self, user):
-        self.session.query(models.Token) \
-            .filter(models.Token.userid == user.userid) \
-            .delete()
+    def _update_tokens(self, old_userid, new_userid):
```

Comment:
It's more concise and performant to just let the database handle most of the work: `self.session.query(models.Token).filter_by(userid=old_userid).update({'userid': new_userid})`

Your test might start breaking because by default SQLAlchemy doesn't synchronize the session for bulk updates. And we don't it to do it in this case. So it's probably easier to just change your test in that case. See [here](http://docs.sqlalchemy.org/en/rel_1_1/orm/query.html?highlight=update#sqlalchemy.orm.query.Query.update) for more information on that.